### PR TITLE
Cs/sc 3360 add road network algorithm

### PR DESCRIPTION
**Pull request by Charl1996**
Same as https://github.com/dimagi/commcare-hq/pull/34081, but with better history.

## Product Description
[Demo video](https://vimeo.com/905617454/0f7b334709?share=copy) (note that the FF was removed)

The disbursement algorithm, "Road Network Algorithm",  makes use of the Mapbox Matrix API. This PR adds the ability for a project to add their own mapbox token on the Geospatial Settings page. 

image

A the new disbursement algorithm will appear on the Geospatial Settings page. Once the algorithm is selected, two new HTML elements will appear on the screen:
1) Text (password) input prompting the user for their mapbox token.
2) A button ("Test API Key") which the user can click to make a sample request to Mapbox in order to verify that the token works.

Note that the user don't have to click the "Test API Key" button in order to save the token; the button exists purely to test the token. 

## Technical Summary
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-3360)

Noteworthy items:
1) The api token is stored in postgres as an encrypted string, but returned to the front end in plaintext (although rendered as a password) when loading the Settings page.
2) The "Test API Key" button sample API call happens purely in the browser; I decided we don't have to have a dedicated endpoint just for the server to make an API call to Mapbox if we could simply use ajax.  

## Feature Flag
Geospatial FF

### Safety story
Tested this feature on my local machine.

### Automated test coverage
- Tests for view
- Tests for model

### QA Plan
QA planned

### Migrations
- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/geospatial/forms.py [hunk 1]
```diff
@@ -73,12 +81,24 @@ class Meta:
...
+        help_text=_(
+            "Enter your Mapbox API token here. Make sure your token has the correct scope configured"
+            " for use of the Mapbox Matrix API."
+        ),
+        required=False,
+    )
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
+        if toggles.SUPPORT_ROAD_NETWORK_DISBURSEMENT_ALGORITHM.enabled(self.domain):
```
Comment (mkangia) [inline corehq/apps/geospatial/forms.py]: Were we able to confirm if we need to add a new feature flag for this?

File: corehq/apps/geospatial/forms.py [hunk 2]
```diff
@@ -195,9 +208,14 @@ def clean(self):
...
-        token = cleaned_data['mapbox_token']
+        algorithm = cleaned_data.get('selected_disbursement_algorithm')
+        token = cleaned_data.get('plaintext_api_token')
         if algorithm == GeoConfig.ROAD_NETWORK_ALGORITHM and not token:
             raise ValidationError(_("Mapbox API token required"))
 
         return cleaned_data
+
+    def save(self, commit=True):
+        if self.cleaned_data.get('plaintext_api_token'):
```
Comment (mkangia) [inline corehq/apps/geospatial/forms.py:214]: I see you changed the name of the property on the form to "plaintext_api_token". Is this if clause still needed?

**Review by mkangia — COMMENTED** [review]
@Charl1996 can you confirm that the commit history is as you expected?
I see a couple of commits that have unrelated changes so confirming once.

File: corehq/apps/geospatial/forms.py (hunk 2 above)
Comment (Charl1996) [reply corehq/apps/geospatial/forms.py]: It's not technically needed, so removing.

File: corehq/apps/geospatial/forms.py (hunk 1 above)
Comment (Charl1996) [reply corehq/apps/geospatial/forms.py]: We don't need it, it just seemed too tedious to remove all references from history for what it's worth, so I took the lazy route and left it as is, since I am removing it [here](https://github.com/dimagi/commcare-hq/pull/34114/commits/8b72bf665ea192904652c52bf986d49a9aa06c7a).

Let me know what you think...if you think "no, let's remove it from history" I'm happy to, otherwise I'll just leave it.

**Review by zandre-eng — APPROVED** [review]
Looks good from my end. Just a few non-blocking comments. Great work on this @Charl1996!

File: corehq/apps/geospatial/forms.py (hunk 1 above)
Comment (mkangia) [reply corehq/apps/geospatial/forms.py]: No worries. I assumed that with the commit clean up, this would not get re-introduced.

**Review by mkangia — APPROVED** [review]
Reviewed all changes instead of going by commit. :+1:

File: corehq/apps/geospatial/views.py [hunk 3]
```diff
@@ -188,6 +191,15 @@ def config(self):
             obj.domain = self.domain
         return obj
 
+    @property
+    def config_dict(self):
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/geospatial/views.py]: https://github.com/dimagi/commcare-hq/pull/34114/commits/7578d4c631fed8f7fe9bbbc0d673ef44568c51f6

**Comment by Charl1996** [comment]
Blocking this PR on the results of [this ticket](https://dimagi.atlassian.net/browse/SC-3447).

File: corehq/apps/geospatial/forms.py (hunk 1 above)
Comment (Charl1996) [reply corehq/apps/geospatial/forms.py]: Just commenting here that [a new FF is indeed now needed](https://dimagi.atlassian.net/browse/SC-3486).

Comment (mkangia) [reply corehq/apps/geospatial/forms.py]: that was quite a ride. This feature flag made it way into the code.

**Comment by Charl1996** [comment]
Closing once again in favour of https://github.com/dimagi/commcare-hq/pull/34199
